src/persist-client/src/internal/service.rs

```diff
@@ -0,0 +1,226 @@
+use std::time::{Instant, SystemTime};
+
+use async_trait::async_trait;
+use futures::Stream;
+use mz_ore::cast::CastFrom;
+use mz_ore::now::SYSTEM_TIME;
+use prost::Message;
+use tokio::sync::mpsc::UnboundedSender;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tokio_stream::StreamExt;
+use tonic::metadata::AsciiMetadataKey;
+use tonic::{Request, Response, Status, Streaming};
+use tracing::{debug, info, info_span, warn};
+
+use crate::internal::metrics::PubSubServerMetrics;
+use mz_persist::location::VersionedData;
+use mz_proto::{ProtoType, RustType};
+
+use crate::rpc::PERSIST_PUBSUB_CALLER_KEY;
+
+include!(concat!(
+    env!("OUT_DIR"),
+    "/mz_persist_client.internal.service.rs"
+));
+
+#[derive(Debug)]
+pub struct PersistService {
+    /// Assigns a unique ID to each incoming connection.
+    connection_id_counter: AtomicUsize,
+    /// Maintains a mapping of `ShardId --> [ConnectionId -> Tx]`.
+    shard_subscribers: Arc<
+        RwLock<
+            BTreeMap<String, BTreeMap<usize, UnboundedSender<Result<ProtoPubSubMessage, Status>>>>,
+        >,
+    >,
+    /// PubSub server-side metrics.
+    metrics: Arc<PubSubServerMetrics>,
+}
+
+impl PersistService {
+    pub fn new(metrics: PubSubServerMetrics) -> Self {
+        PersistService {
+            shard_subscribers: Default::default(),
+            connection_id_counter: AtomicUsize::new(0),
+            metrics: Arc::new(metrics),
+        }
+    }
+}
+
+#[async_trait]
+impl proto_persist_pub_sub_server::ProtoPersistPubSub for PersistService {
+    type PubSubStream = Pin<Box<dyn Stream<Item = Result<ProtoPubSubMessage, Status>> + Send>>;
+
+    async fn pub_sub(
+        &self,
+        req: Request<Streaming<ProtoPubSubMessage>>,
+    ) -> Result<Response<Self::PubSubStream>, Status> {
+        let root_span = info_span!("persist::push::server");
+        let _guard = root_span.enter();
+        let caller_id = req
+            .metadata()
+            .get(AsciiMetadataKey::from_static(PERSIST_PUBSUB_CALLER_KEY))
+            .map(|key| key.to_str().ok())
+            .flatten()
+            .map(|key| key.to_string())
+            .unwrap_or_else(|| "unknown".to_string());
+        info!("incoming push from: {:?}", caller_id);
+
+        let mut in_stream = req.into_inner();
+
+        // WIP not unbounded
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+        let connection_id = self.connection_id_counter.fetch_add(1, Ordering::SeqCst);
+
+        let subscribers = Arc::clone(&self.shard_subscribers);
+        // this spawn here to cleanup after connection error / disconnect, otherwise the stream
+        // would not be polled after the connection drops. in our case, we want to clear the
+        // connection and its subscriptions from our shared state when it drops.
+        let metrics = Arc::clone(&self.metrics);
+        tokio::spawn(async move {
+            let root_span = info_span!("persist::push::server_conn");
```

Comment:
feels like this whole task impl should be moved into its own method. probably ditto for the impls of each message type?